library/ssl_tls13_generic.c

```diff
@@ -136,6 +136,84 @@ void mbedtls_ssl_tls13_add_hs_hdr_to_checksum( mbedtls_ssl_context *ssl,
 }
 
 #if defined(MBEDTLS_KEY_EXCHANGE_WITH_CERT_ENABLED)
+/* mbedtls_ssl_tls13_parse_sig_alg_ext()
+ *
+ * enum {
+ *    ....
+ *   ecdsa_secp256r1_sha256( 0x0403 ),
+ *   ecdsa_secp384r1_sha384( 0x0503 ),
+ *   ecdsa_secp521r1_sha512( 0x0603 ),
+ *    ....
+ * } SignatureScheme;
+ *
+ * struct {
+ *    SignatureScheme supported_signature_algorithms<2..2^16-2>;
+ * } SignatureSchemeList;
+ */
+int mbedtls_ssl_tls13_parse_sig_alg_ext( mbedtls_ssl_context *ssl,
+                                         const unsigned char *buf,
+                                         const unsigned char *end )
+{
+    const unsigned char *p = buf;
+    size_t sig_algs_len = 0;
+    uint16_t *sig_algs;
+    const unsigned char *sig_algs_end;
+    uint16_t sig_alg;
+    uint32_t common_idx = 0; /* iterate through received signature schemes list */
+
+    /* skip 2 bytes of signature algorithms length */
+    MBEDTLS_SSL_CHK_BUF_READ_PTR( p, end, 2 );
+    sig_algs_len = MBEDTLS_GET_UINT16_BE( p, 0 );
+    p += 2;
+
+    sig_algs = mbedtls_calloc( MBEDTLS_SIG_ALGS_SIZE, sizeof( uint16_t ) );
```

Comment:
```suggestion
    ssl->handshake->received_signature_schemes_list = mbedtls_calloc( 1, received_sig_algs_len );
```
count the length in bytes

Reply:
Please do not allocate this. Let's do it as in the prototype: store the received list of algorithms in an array statically allocated in the handshake context: a field `received_sig_algs[ MBEDTLS_RECEIVED_SIG_ALGS_SIZE ];` should make it.